sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/JacksonParser.scala

```diff
@@ -35,184 +34,303 @@ import org.apache.spark.util.Utils
+          } else {
+            // In Spark 1.5.0, we store the data as number of days since epoch in string.
+            // So, we just convert it to Int.
+            stringValue.toInt
+          }
+      }
 
-      case (VALUE_NUMBER_INT, ShortType) =>
-        parser.getShortValue
+    case BinaryType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case VALUE_STRING => parser.getBinaryValue
+      }
 
-      case (VALUE_NUMBER_INT, IntegerType) =>
-        parser.getIntValue
+    case dt: DecimalType =>
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case (VALUE_NUMBER_INT | VALUE_NUMBER_FLOAT) =>
+          Decimal(parser.getDecimalValue, dt.precision, dt.scale)
+      }
 
-      case (VALUE_NUMBER_INT, LongType) =>
-        parser.getLongValue
+    case st: StructType =>
+      val fieldConverters = st.map(_.dataType).map(makeConverter)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertObject(parser, st, fieldConverters)
+      }
 
-      case (VALUE_TRUE, BooleanType) =>
-        true
+    case at: ArrayType =>
+      val elementConverter = makeConverter(at.elementType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_ARRAY => convertArray(parser, elementConverter)
+      }
 
-      case (VALUE_FALSE, BooleanType) =>
-        false
+    case mt: MapType =>
+      val valueConverter = makeConverter(mt.valueType)
+      (parser: JsonParser) => parseJsonToken(parser, dataType) {
+        case START_OBJECT => convertMap(parser, valueConverter)
+      }
 
-      case (START_OBJECT, st: StructType) =>
-        convertObject(factory, parser, st)
+    case udt: UserDefinedType[_] =>
+      makeConverter(udt.sqlType)
 
-      case (START_ARRAY, ArrayType(st, _)) =>
-        convertArray(factory, parser, st)
+    case _ =>
+      (parser: JsonParser) =>
+        // Here, we pass empty `PartialFunction` so that this case can be
+        // handled as a failed conversion. It will throw an exception as
+        // long as the value is not null.
+        parseJsonToken(parser, dataType)(PartialFunction.empty[JsonToken, Any])
+  }
 
-      case (START_OBJECT, MapType(StringType, kt, _)) =>
-        convertMap(factory, parser, kt)
+  /**
+   * This handles nulls ahead before trying to check the tokens, and applies the conversion
+   * function and then checks failed the conversion afterward if the `f` fails to convert
+   * the value.
+   *
+   * In more details, it checks `FIELD_NAME` if exists and then skip. If this is called after
+   * `START_OBJECT`, then, the next token can be `FIELD_NAME`. Since the names are kept in
+   * `JacksonParser.convertObject`, this `FIELD_NAME` token can be skipped as below. When this
+   * is called after `START_ARRAY`, the tokens are consecutive tokens for values without
+   * `FIELD_NAME` until `END_ARRAY`. In this case, we don't have to skip.
```

Comment:
it's good to know this, but do we really need to put it in comments? It seems not that helpful for people to understand the code. We just need some comments `case FIELD_NAME =>` to explain why we need to skip, i.e. there are useless `FIELD_NAME`s between `START_OBJECT` and `END_OBJECT` tokens.